README.md
# Django Ajax View
Simple Django library for quickly using Ajax calls in regular views.

For example, suppose you have a regular TemplateView in Django, which you want to add Ajax interactivity to. Simply add an `@ajax` method to your controller, and it becomes accessible in the HTML template side as a function:

```python
class MyController(AjaxView):
    ...

    @ajax
    def greetings(self, name):
        return f'Why hello there, {name}!'
```

Inside the rendered HTML:

```javascript
const message = await greetings({name: 'Ken'})
```

See [documentation for details](https://django-ajaxview.readthedocs.io/). Licensed under MIT.


docs/changelog.md
# Change Log

## Backwards Compatibility

Minor versions are intended to be backwards compatible. For example, `1.9.0` would be backwards compatible to `1.0.0', but `2.0.0' would not be backwards compatible to `1.9.0'.

## Release Log

### 2020-12-13: 1.0.1

First public release.

### 2020-12-26: 1.0.3

Fixed CI. This was the first version uploaded to PyPi automatically. No functional changes.

### 2021-10-11: 1.0.4

Fixed no argument functions causing errors

### 2022-01-13: 1.0.6

Fixed base classes not working


test_settings.py
"""
Django settings for AjaxViewDemo project.

Generated by 'django-admin startproject' using Django 3.1.4.

For more information on this file, see
https://docs.djangoproject.com/en/3.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.1/ref/settings/
"""

from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent
SECRET_KEY = 'this-is-test-data'
DEBUG = True

ALLOWED_HOSTS = []

# Application definition

INSTALLED_APPS = [
    'ajaxview'
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'AjaxViewDemo.wsgi.application'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': str(BASE_DIR / 'db.sqlite3'),
    }
}
LANGUAGE_CODE = 'en-us'
TIME_ZONE = 'UTC'
USE_I18N = True
USE_L10N = True
USE_TZ = True


ajaxview/tests/test_ajaxview.py
import json
import logging
from datetime import datetime, timedelta

from django.test import TestCase, RequestFactory


class TestAjaxView(TestCase):
    def setUp(self):
        # Every test needs access to the request factory.
        self.factory = RequestFactory()

    def get_controller(self, suppress_logging=False):
        from ajaxview import AjaxView, ajax

        # Avoid printing traceback to log, since it's expected
        import logging
        if suppress_logging:
            logger = logging.Logger('test')
            logger.setLevel(100)
        else:
            logger = logging

        class MyAjaxController(AjaxView):
            def get_logger(self):
                return self.logger

            @ajax
            def sum_numbers(self, numbers):
                return sum(numbers)

            @ajax
            def sum_numbers_server_error(self, numbers):
                raise ValueError('I suck')

            @ajax
            def sum_numbers_client_error(self, numbers):
                raise self.ClientError('You suck')

            @ajax
            def add_days(self, when: datetime, days):
                return when + timedelta(days=days)

        MyAjaxController.logger = logger
        return MyAjaxController

    def test_rendering(self):
        MyAjaxController = self.get_controller()

        ct = MyAjaxController(request=self.factory.get('/test'))
        self.assertTrue(ct.get_context_data()['ajax_view'].strip().startswith('<script>'))

    def test_working_function(self):
        """
        Just tests that the working function is there
        """
        self.assert_call_result('sum_numbers', {'numbers': [1, 2, 3]}, 200, {
            'response_type': 'complete',
            'return_value': 6
        })

    def test_bad_request_missing_args(self):
        MyAjaxController = self.get_controller()

        ct = MyAjaxController(request=self.factory.get('/test'))
        request = self.factory.post('/test', {
            'func': 'sum_numbers',
        }, content_type='application/json')
        response = ct.post(request=request)
        data = json.loads(response.content)
        self.assertEqual(response.status_code, 400)
        self.assertEqual(data['response_type'], 'invalid-request')
        self.assertEqual(data['errors'], ['No args specified'])

        request = self.factory.post('/test', {
            'args': {'numbers': [1, 2, 3]}
        }, content_type='application/json')

        response = ct.post(request=request)
        data = json.loads(response.content)
        self.assertEqual(response.status_code, 400)
        self.assertEqual(data['response_type'], 'invalid-request')
        self.assertEqual(data['errors'], ['No func specified'])

    def test_client_error(self):
        self.assert_call_result('sum_numbers_client_error', {'numbers': [1, 2, 3]}, 400, {
            'response_type': 'client-error',
            'errors': ['You suck']
        })

    def test_server_error(self):
        self.assert_call_result('sum_numbers_server_error', {'numbers': [1, 2, 3]}, 500, {
            'response_type': 'server-error'
        }, suppress_logging=True)

    def test_no_such_function(self):
        self.assert_call_result('does_not_exist', {'foo': 'bar'}, 400, {
            'response_type': 'invalid-request',
            'errors': ['No such ajax function: does_not_exist']
        })

    def test_positional_args(self):
        self.assert_call_result('sum_numbers', [1, 2, 3], 400, {
            'response_type': 'invalid-request',
            'errors': ['args must be dict']
        })

    def test_not_json(self):
        MyAjaxController = self.get_controller()

        request = self.factory.post('/test', 'your mom uses Linux', content_type='application/json')
        ct = MyAjaxController(request=request)
        response = ct.post(request=request)
        data = json.loads(response.content)

        self.assertEqual(response.status_code, 400)
        self.assertEqual(data['response_type'], 'invalid-request')
        self.assertEqual(data['errors'], ['Cannot parse JSON body'])

    def test_datetime_serialization(self):
        self.assert_call_result('add_days', {'when': '2020-12-14T00:45:08.651Z', 'days': 5}, 200, {
            'response_type': 'complete',
            'return_value': '2020-12-19T00:45:08.651'
        })

    def assert_call_result(self, function_name, function_args, http_status, data_values, suppress_logging=False):
        MyAjaxController = self.get_controller(suppress_logging=suppress_logging)

        request = self.factory.post('/test', {
            'args': function_args,
            'func': function_name,
        }, content_type='application/json')

        ct = MyAjaxController(request=request)
        response = ct.post(request=request)
        data = json.loads(response.content)

        self.assertEqual(response.status_code, http_status)
        for k, v in data_values.items():
            self.assertEqual(data[k], v)

    def test_get_logging(self):
        from ajaxview import AjaxView

        class MyAjaxController(AjaxView):
            pass

        ct = MyAjaxController()
        logger = ct.get_logger()
        assert logger is logging


ajaxview/__init__.py
import inspect
import json
import logging
from datetime import datetime

from django.http import JsonResponse, HttpRequest
from django.template.context_processors import csrf
from django.template.loader import render_to_string
from django.views.generic import TemplateView

from . import exceptions


def ajax(fn):
    fn.is_ajax = True
    return fn


class AjaxViewMeta(type):
    def __new__(cls, name, bases, dct):
        class_instance = super().__new__(cls, name, bases, dct)

        ajax_methods = {}
        ajax_methods_js = {}  # Handy for rendering in templates
        ajax_datetime_params = {}

        # Load stuff from super-classes
        for base_class in bases:
            if hasattr(base_class, 'ajax_methods'):
                ajax_methods.update(base_class.ajax_methods)
            if hasattr(base_class, 'ajax_methods_js'):
                ajax_methods_js.update(base_class.ajax_methods_js)
            if hasattr(base_class, 'ajax_datetime_params'):
                ajax_datetime_params.update(base_class.ajax_datetime_params)

        for k, v in dct.items():
            if callable(v) and getattr(v, 'is_ajax', False):
                sig = inspect.signature(v)
                args = list(sig.parameters.keys())
                if args[0] == 'self' or args[0] == 'cls':
                    args = args[1:]
                ajax_methods[k] = args
                ajax_datetime_params[k] = []
                ajax_methods_js[k] = ','.join(args)
                for arg, param in sig.parameters.items():
                    if param.annotation is datetime:
                        ajax_datetime_params[k].append(arg)

        class_instance.ajax_methods = ajax_methods
        class_instance.ajax_methods_js = ajax_methods_js
        class_instance.ajax_datetime_params = ajax_datetime_params
        return class_instance


class AjaxView(TemplateView, metaclass=AjaxViewMeta):
    ajax_function_name = 'ajax'
    ClientError = exceptions.ClientError

    def get_logger(self):
        return logging

    def get_context_data(self, **kwargs):
        kwargs.setdefault('ajax_view', self.render_page_head())
        return super().get_context_data(**kwargs)

    def render_bad_request(self, error):
        return JsonResponse({
            'response_type': 'invalid-request',
            'errors': [error]
        }, status=400)

    def render_client_error(self, exception_instance):
        error = exception_instance.args

        return JsonResponse({
            'response_type': 'client-error',
            'errors': list(error)
        }, status=400)

    def render_server_error(self, exception_instance):
        return JsonResponse({
            'response_type': 'server-error',
        }, status=500)

    def render_user_response(self, return_value):
        return JsonResponse(data={
            'response_type': 'complete',
            'return_value': return_value
        })

    def render_page_head(self):
        context = dict(
            csrf_token=csrf(self.request)['csrf_token'],
            ajax_function_name=self.ajax_function_name,
            methods=self.ajax_methods_js
        )

        return render_to_string('ajax_headers.html', context)

    def post(self, request: HttpRequest, *args, **kwargs):
        try:
            input_data = json.loads(request.body)
        except:
            return self.render_bad_request('Cannot parse JSON body')

        func = input_data.get('func', None)
        arguments = input_data.get('args', None)

        if not func:
            return self.render_bad_request('No func specified')

        if arguments is None:
            return self.render_bad_request('No args specified')

        if not isinstance(arguments, dict):
            return self.render_bad_request('args must be dict')

        # Convert any datetimes
        if func in self.ajax_datetime_params:
            for k, v in arguments.copy().items():
                if k in self.ajax_datetime_params[func]:
                    arguments[k] = datetime.strptime(v, '%Y-%m-%dT%H:%M:%S.%fZ')

        if func in self.ajax_methods:
            try:
                data = getattr(self, func)(**arguments)
            except self.ClientError as e:
                return self.render_client_error(e)
            except Exception as e:
                self.get_logger().exception(f'Error processing {func}')
                return self.render_server_error(e)
            else:
                return self.render_user_response(data)
        else:
            return self.render_bad_request(f'No such ajax function: {func}')

__all__ = ['AjaxView', 'ajax']
